db/db_impl_compaction_flush.cc

```diff
@@ -287,6 +287,27 @@ Status DBImpl::CompactRange(const CompactRangeOptions& options,
   auto cfd = cfh->cfd();
   bool exclusive = options.exclusive_manual_compaction;
 
+  if (!options.allow_write_stall) {
+    InstrumentedMutexLock l(&mutex_);
+    WriteStallCondition write_stall_condition = WriteStallCondition::kNormal;
+    do {
+      if (write_stall_condition != WriteStallCondition::kNormal) {
+        TEST_SYNC_POINT("DBImpl::CompactRange:StallWait");
+        bg_cv_.Wait();
+      }
+      const auto& mutable_cf_options = *cfd->GetLatestMutableCFOptions();
+      const auto* vstorage = cfd->current()->storage_info();
+      // check whether one extra immutable memtable or an extra L0 file would
+      // cause write stalling mode to be entered. It could still enter stall
+      // mode due to pending compaction bytes, but that's less common
+      write_stall_condition =
+          ColumnFamilyData::GetWriteStallConditionAndCause(
+              cfd->imm()->NumNotFlushed() + 1,
+              vstorage->l0_delay_trigger_count() + 1,
+              vstorage->estimated_compaction_needed_bytes(), mutable_cf_options)
+              .first;
+    } while (write_stall_condition != WriteStallCondition::kNormal);
```

Comment:
There may be a chance that because L0->L1 keeps being kicked out so CompactRange() will stuck here forever.

One thing we can do is that after https://github.com/facebook/rocksdb/pull/3378 is checked in, we can check whether the current memtable has already been flushed by automatic flush, using the technique used there. If it happens, we can skip the flush step and just continue the next step.

Reply:
Actually, I didn't fully understand the case of being stuck here forever. What does L0->L1 being kicked out mean?

Reply:
I mean after L0->L1 finishes, number of L0 files have been increase so much that stalling condition happens again. So L0->L1 keeps going forever and number of L0 files never drops to a number that allows the compaction to run.

Reply:
Got it, thanks.